src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -802,27 +802,22 @@ public override TypeSymbolWithAnnotations AsNotNullableReferenceType()
 
             public override TypeSymbolWithAnnotations SubstituteType(AbstractTypeMap typeMap)
             {
-                if ((object)_resolved != null)
+                var newUnderlying = _underlying.SubstituteType(typeMap);
+                if (newUnderlying.Equals(_underlying, TypeCompareKind.AllAspects))
```

Comment:
>if (newUnderlying.Equals(_underlying, TypeCompareKind.AllAspects)) [](start = 16, length = 66)

The implementation doesn't look equivalent. In particular, it doesn't look like it will properly handle presence of custom modifiers after the substitution.